Issue/8670 site creation errors segments

malinajirka opened the pull request:
Fixes #8670 


1. Removes slide-in animation for the first screen of the Site Creation flow

2. UiState refactoring - the structure we came up with felt like an overkill and just complicated things. I also changed my mind regarding having ideally no conditional statements in view layer. As long as they are straightforward I think it's ok to have them. So if you don't mind I decided to go with simpler approach -> I've separated the UiState into two sub states -> Content/Error -> if a value in the UiState is null ->  we hide the view. I'm sorry for keep doing these changes, but I realized how complicated and not-flexible it is while I was introducing the error states.

3. Introduces TEST_DISPATCHER which doesn't wait on `delay(n)` calls.

4. Adds support for "No Connection" error

To test:
- set the `NEW_SITE_CREATION_ENABLED` build time feature flag to `true`
1. Start the SiteCreation flow
2. Wait until segments loads
3. Select a segment
4. Make sure you are navigated to the Verticals screen
-------
1. Turn on Airplane mode
2. Start the SiteCreation flow
3. Notice the "No connection" error message
4. Click on Retry
5. Notice the "No connection" error message
6. Turn off airplane mode
7. Click on Retry
8. Make sure the segments are shown
--------
- add `event.error = FetchSegmentsError(GENERIC_ERROR)` to `FetchSegmentsUseCase.onSiteCategoriesFetched(...)`
1. Start the SiteCreation flow
2. Notice the "There was a problem" error
--------
![untitled](https://user-images.githubusercontent.com/2261188/49507832-950e1d80-f881-11e8-9354-0dc9baf17110.png)

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsViewModel.kt:
```diff
@@ -14,21 +15,24 @@ import org.wordpress.android.fluxc.Dispatcher
 import javax.inject.Named
 import kotlin.coroutines.experimental.CoroutineContext
 
+private const val FAKE_DELAY = 1000
```
`FAKE_DELAY` is not a very descriptive name, can we improve this? Any name that describes what this is actually used for (however long) should be fine. Also, I know you have a comment for this in its usage, but I think it's worth to add a comment here if we can't find a descriptive enough name for it.

malinajirka replied:
I've renamed it to`SHOW_LOADING_TO_PROVIDE_FEEDBACK_ARTIFICIAL_DELAY` let me know if you come up with better name;).

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/NetworkUtilsWrapper.kt:
```diff
@@ -0,0 +1,13 @@
+import javax.inject.Singleton
+
+@Singleton
+class NetworkUtilsWrapper @Inject constructor() {
```
Let's pick an approach between this approach and [ConnectionStatusLiveData](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/viewmodel/helpers/ConnectionStatusLiveData.kt) and use it everywhere. It's not great to add 2 different components that achieves the same thing immediately after one another.

I am fine with either approach, so feel free to go with whichever you prefer, but if you do want to use this one, it'd be great if you can open a PR to remove the `ConnectionStatusLiveData`.

Also, can you please update the base branch with `develop`? Our tests and reviews would be more reliable if we can keep up with the base branch. If you go with the above change, it'd be necessary to do so anyway.

malinajirka replied:
>>Let's pick an approach between this approach and ConnectionStatusLiveData and use it everywhere. It's not great to add 2 different components that achieves the same thing immediately after one another.

>Hmm, I don't consider them to be the same components. One registers a broadcast receiver which can be a bit of an overkill and the second one just asks on the current status. However, I'm fine with keeping just a single component. The ConnectionStatusLiveData is more powerful so it's an easy choice.

Ok, so it's not so easy choice. I think we should keep both of them. IMHO they serve a different purpose.
The `ConnectionStatusLiveData` needs to be observed in order to register to the Broadcast -> we'd have to start observing in the view layer and propagate the changes to the VM. However, since we want to start fetching data in `vm.onStart()` we can't be sure the broadcast reciever is already active and we'll receive the result in time. Moreover, it feels like we are using something aimed for a different purpose.... I think the `ConnectionStatusLiveData` is useful for cases where we want to for example automatically update the UI when the connection status changes. Which is not the case. Wdyt?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsViewModel.kt:
```diff
@@ -154,30 +176,37 @@ class NewSiteCreationSegmentsViewModel
-        val showContent: Boolean,
-        val items: List<ItemUiState>
-    )
+interface SegmentsUiState
```
This should be a `sealed class` since there are only 2 possible values for it and they are mutually exclusive.

oguzkocer replied:
With my other comments combined, here is the approach I'd have taken:

```
sealed class SegmentsUiState {
    data class SegmentsContentUiState(val items: List<SegmentsItemUiState>) : SegmentsUiState()
    sealed class SegmentsErrorUiState constructor(
        val titleResId: Int,
        val subtitleResId: Int? = null
    ) : SegmentsUiState() {
        object genericErrorUiState : SegmentsErrorUiState(
                R.string.site_creation_error_generic_title,
                R.string.site_creation_error_generic_subtitle
        )

        object connectionErrorUiState : SegmentsErrorUiState(R.string.site_creation_error_connection_title)
    }
}
```

malinajirka replied:
👍

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsViewModel.kt:
```diff
@@ -154,30 +176,37 @@ class NewSiteCreationSegmentsViewModel
-            val titleResId: Int = R.string.site_creation_segments_title
-            val subtitleResId: Int = R.string.site_creation_segments_subtitle
-        }
+        fun createConnectionErrorUiState() = SegmentsErrorUiState(R.string.site_creation_error_connection_title)
```
Can we change this to a `val`?

```suggestion
        val createConnectionErrorUiState = SegmentsErrorUiState(R.string.site_creation_error_connection_title)
```

malinajirka replied:
👍 -> I've used the proposed solution from the other comment.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsViewModel.kt:
```diff
@@ -154,30 +176,37 @@ class NewSiteCreationSegmentsViewModel
+data class SegmentsContentUiState(val items: List<SegmentsItemUiState>) : SegmentsUiState
+data class SegmentsErrorUiState constructor(val titleResId: Int, val subtitleResId: Int? = null) : SegmentsUiState {
+    companion object {
+        fun createGenericErrorUiState() = SegmentsErrorUiState(
```
Can we change this to a `val`?

```suggestion
        val genericErrorUiState = SegmentsErrorUiState(
```

malinajirka replied:
👍 -> I've used the proposed solution from the other comment.

oguzkocer commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -2352,4 +2352,7 @@
     <string name="new_site_creation_button_skip">Skip</string>
+    <string name="site_creation_error_generic_title">There was a problem</string>
+    <string name="site_creation_error_generic_subtitle">Error communicating with the server, please try again</string>
+    <string name="site_creation_error_connection_title">No internet connection</string>
```
Is there a reason we are not using the `no_network_message` resource instead?

malinajirka replied:
@SylvesterWilmott  Could we reuse following resource "There is no network available" on the fullscreen error screen?

oguzkocer commented on WordPress/src/test/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsViewModelTest.kt:
```diff
@@ -57,10 +59,10 @@ private val SECOND_MODEL =
-private val HEADER_AND_FIRST_ITEM_STATE = UiState(
-        false, true,
+private val PROGRESS_STATE = SegmentsContentUiState(listOf(HeaderUiState, ProgressUiState))
+private val GENERIC_ERROR_STATE = SegmentsErrorUiState.SegmentsGenericErrorUiState
```
I am not sure if there is any value having these values anymore. Can we just use `SegmentsGenericErrorUiState` in place of `GENERIC_ERROR_STATE` since it's an object? Also, same thing for `NO_CONNECTION_ERROR_STATE`.

P.S: The only value I see for keeping it this way is to avoid having to change it if we need to make a change to the structure. In that case, I think it's better to force ourselves to make the change in the tests explicitly to make sure we think about its implications. Also, as you know by now, I am not a fan of having unused code or code that might be useful in the future in a code base.

malinajirka replied:
I thought it'd be more readable if we kept everything written in upper_case. However, it doesn't really matter. I've removed them;).

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsViewModel.kt:
```diff
@@ -14,21 +15,26 @@ import org.wordpress.android.fluxc.Dispatcher
 import javax.inject.Named
 import kotlin.coroutines.experimental.CoroutineContext
 
+private const val SHOW_LOADING_TO_PROVIDE_FEEDBACK_ARTIFICIAL_DELAY = 1000
```
Maybe `CONNECTION_ERROR_DELAY_TO_SHOW_LOADING_STATE_FOR_FEEDBACK`, if it's too long `CONNECTION_ERROR_DELAY_TO_SHOW_LOADING_STATE` could also work.

I am totally fine with the current one as well, it's much better than fake delay 🙈

malinajirka replied:
👍

malinajirka commented:
Thanks @oguzkocer! I've renamed the "fake_delay" variable and I've removed the constants for the errors.

[This](https://github.com/wordpress-mobile/WordPress-Android/pull/8732/files#diff-e0a4f5dcdd4015ad495450cbdbafd431) change in `NewSiteCreationVerticalsViewModelTest.kt` should fix the master branch. I'm sorry for the broken branch 😞

I'd suggest leaving the last two issues (ConnectionStatusLiveData and no_network_message) for another PR. However, we can continue discussing it here so the whole conversation is at a single place. Wdyt?

oguzkocer commented:
As discussed on Slack, we are merging this before the design review pass since it's targeting a WIP branch so we can fix the base branch after open PRs are merged. @malinajirka will follow up with @SylvesterWilmott for the design review.

:shipit:
